h2o-algos/src/main/java/hex/gam/GAMModel.java

```diff
@@ -0,0 +1,437 @@
+    String[] testNames = test.names();
+    if (!equalColNames(testNames, _output._dinfo._adaptedFrame.names(), _parms._response_column)) {  // shallow check: column number, column names only
+      Frame adptedF = cleanUpInputFrame(test, testNames);
+      int testNumCols = test.numCols();
+      for (int index=0; index < testNumCols; index++)
+        test.remove(0);
+      int adaptNumCols = adptedF.numCols();
+      for (int index=0; index < adaptNumCols; index++)
+        test.add(adptedF.name(index), adptedF.vec(index));
+
+      return super.adaptTestForTrain(adptedF, expensive, computeMetrics);
+    }
+    return super.adaptTestForTrain(test, expensive, computeMetrics);
+  }
+  
+  public Frame cleanUpInputFrame(Frame test, String[] testNames) {
+    Frame adptedF = new Frame(Key.make(), test._names.clone(), test.vecs().clone()); // clone test dataset
+    int numGamCols = _output._numKnots.length;
+    Vec[] gamCols = new Vec[numGamCols];
+    for (int vind=0; vind<numGamCols; vind++)
+      gamCols[vind] = adptedF.vec(_parms._gam_X[vind]).clone();
+    Frame onlyGamCols = new Frame(_parms._gam_X, gamCols);
+    AddGamColumns genGamCols = new AddGamColumns(_output._binvD, _output._knots, _output._numKnots, _output._dinfo, 
+            onlyGamCols);
+    genGamCols.doAll(genGamCols._gamCols2Add, Vec.T_NUM, onlyGamCols);
+    String[] gamColsNames = new String[genGamCols._gamCols2Add];
+    int offset = 0;
+    for (int ind=0; ind<genGamCols._numGAMcols; ind++) {
+      System.arraycopy(_gamColNames[ind], 0, gamColsNames, offset, _gamColNames[ind].length);
+      offset+=_gamColNames[ind].length;
+    }
+    Frame oneAugmentedColumn = genGamCols.outputFrame(Key.make(), gamColsNames, null);
+    if (_parms._ignored_columns != null) {  // remove ignored columns
+      for (String iname:_parms._ignored_columns) {
+        if (ArrayUtils.contains(testNames, iname)) {
+          adptedF.remove(iname);
+        }
+      }
+    }
+    int numCols = adptedF.numCols();  // remove constant or bad frames.
+    for (int vInd=0; vInd<numCols; vInd++) {
+      Vec v = adptedF.vec(vInd);
+      if ((_parms._ignore_const_cols &&  v.isConst()) || v.isBad())
+        adptedF.remove(vInd);
+    }
+    Vec respV = null;
+    if (ArrayUtils.contains(testNames, _parms._response_column))
+      respV = adptedF.remove(_parms._response_column);
+    adptedF.add(oneAugmentedColumn.names(), oneAugmentedColumn.removeAll());
+    Scope.track(oneAugmentedColumn);
+    if (respV != null)
+      adptedF.add(_parms._response_column, respV);
+    return adptedF;
+  }
+  
+  @Override
+  protected Frame predictScoreImpl(Frame fr, Frame adaptFrm, String destination_key, Job j, boolean computeMetrics, 
+                                   CFuncRef customMetricFunc) {
+    int nReponse = ArrayUtils.contains(adaptFrm.names(), _parms._response_column)?1:0;
+    String[] predictNames = super.makeScoringNames();
```

Comment:
scoring here is done with original model matrix and coefficients as described in 7.5